### ISSUE TITLE
catch "when" filter applied for unsupporting language levels

### ISSUE THREAD

**Issue by Niko-O** [icsharpcode/ILSpy#2455]
### Input code
```c#
static void Main(string[] args)
{
    try
    {
    }
    catch (Exception ex) when (ex.Message.Length > 10)
    {
    }
}
```

### Erroneous output
When decompiled with "C# 4.0 / VS 2010" selected, the output looks exactly like the input, but that does not compile in C# 4.0. The error messages when trying to compile that are (translated from german so they might not be exact):
1. "Expected {" at "when"
2. "Expected ;" at the end of the line after "10)"
3. "The Name 'when' is not available in the current context" at "when"

### Details
* Product in use: Cloned the repository in its current state, doubleclicked "ILSpy.sln", Started debugging.
* Version in use: ILSpy version 7.1.0.0-no-branch

**Comment by siegfriedpammer** [comment]
Yes, there is currently no setting that allows disabling decompilation of exception filters. @dgrunwald @christophwille do you think this is worth adding? I am a bit torn as there isn't really a way to represent code that uses exception filters when disabling the feature. However, this could be said about any language feature that isn't just syntactic sugar, so maybe adding a setting for exception filters is the right thing to do.

**Comment by Niko-O** [comment]
A possible workaround would be to generate code like this:
```c#
static void Main(string[] args)
{
    try
    {
    }
    catch (Exception ex)
    {
        if (!(ex.Message.Length > 10))
        {
            // This was part of a filter block on the current exception; when-filters not available in the selected language level.
            throw;
        }
    }
}
```
Possibly less verbose than that.

**Comment by christophwille** [comment]
What is our "default" strategy for decompiling a language feature that is net-new in a higher language version, but the user has selected an older version that can't represent that feature correctly?

**Comment by dgrunwald** [comment]
Note that catch-rethrow has different semantics than an exception filter -- the stack trace will look different, and the evaluation order with regards to `finally` blocks is different.
So in general it's not correct to perform the substitution you suggest.

Our usual strategy that is that we only offer options for C# language features that are effectively syntax sugar. The C# features that allow access to IL constructs previously inaccessible (e.g. exception filters, ref locals, function cannot be compiled down to older language versions and will instead always use the new language feature.

However, there is some precedent for intentionally decompiling to incorrect code if the correct code cannot be represented in the chosen language version and the difference in semantics is unlikely to matter: bug #2050 is only fixed for C# >= 6 and intentionally not fixed for lower language versions.

**Comment by Niko-O** [comment]
> the stack trace will look different

That is not the case as far as I know the purpose of `throw;` (note: **not** `throw ex;`!). The original StackTrace will be kept intact.

> and the evaluation order with regards to finally blocks is different.

As far as my test goes, the behavior is the same:
```c#
static void Main(string[] args)
{
    Console.WriteLine("A");
    try
    {
        Console.WriteLine("B");
        DeepInTheBlues1();
        Console.WriteLine("J");
    }
    catch (Exception ex)
    {
        Console.WriteLine("K");
        Console.WriteLine(ex.Message);
        Console.WriteLine(ex.StackTrace);
    }
    finally
    {
        Console.WriteLine("L");
    }
    Console.WriteLine("M");
    Console.ReadLine();
}

static void DeepInTheBlues1()
{
    Console.WriteLine("C");
    try
    {
        Console.WriteLine("D");
        DeepInTheBlues2();
        Console.WriteLine("E");
    }
    catch (Exception ex)
#if false
    when (new Func<bool>(() =>
            {
        Console.WriteLine("F");
        return ex.Message.Length > 5;
    }).Invoke())
#endif
        {
#if true
        Console.WriteLine("F");
        if (!(ex.Message.Length > 5)) throw;
#endif
        Console.WriteLine("G");
        Console.WriteLine(ex.Message);
        Console.WriteLine(ex.StackTrace);
    }
    finally
    {
        Console.WriteLine("H");
    }
    Console.WriteLine("I");
}

static void DeepInTheBlues2()
{
    DeepInTheBlues3();
}
static void DeepInTheBlues3()
{
    throw new Exception("12345678");
    //throw new Exception("1234");
}
```
(Comment and un-comment accordingly.)
![Diff](https://user-images.githubusercontent.com/4943909/126480689-a1a79264-71ba-4aca-bc87-92bc41c0bd2d.png)
Are you referring to more subtle behavior that's not visible in my test?

> The C# features that allow access to IL constructs previously inaccessible [...] will instead always use the new language feature.

Seems like a reasonable choice. It's difficult to strike a balance between "code that does the same thing in the end" and "code that accurately represents what was originally there in IL".

**Comment by dgrunwald** [comment]
Add a try-finally block inside `DeepInTheBlues2`.
I believe the `when` clause will be evaluated before the stack is unwound = `when` in `DeepInTheBlues1` runs before `finally` in `DeepInTheBlues2`

**Comment by Niko-O** [comment]
Whoa! That's unexpected! I tried it and yes, you are right!
That's so weird... I wonder how that works. Because you can access local variables inside the `when`.
Well, I'd say in this case it's OK to just decompile to code that uses `when` and let the human decide what to do with that code.
It would be awesome to still make that an option, but with a warning about that order-of-execution-problem next to it. For example, in my case I have to fix a bunch of those filters manually and I know that the order won't make a difference (and I'd say it doesn't in most cases).
